be/src/olap/rowset/beta_rowset_writer.cpp

```diff
@@ -102,6 +115,409 @@ Status BetaRowsetWriter::add_block(const vectorized::Block* block) {
+        LOG(WARNING) << "failed to get segcompaction writer";
+        return Status::OLAPInternalError(OLAP_ERR_SEGCOMPACTION_INIT_WRITER);
+    }
+    uint64_t row_count = 0;
+    vectorized::Block block = _context.tablet_schema->create_block();
+    while (true) {
+        auto status = reader_ptr->next_batch(&block);
+        row_count += block.rows();
+        if (status != Status::OK()) {
+            if (LIKELY(status.is_end_of_file())) {
+                RETURN_NOT_OK_LOG(_add_block_for_segcompaction(&block, &writer),
+                                  "write block failed");
+                break;
+            } else {
+                LOG(WARNING) << "read block failed: " << status.to_string();
+                return status;
+            }
+        }
+        RETURN_NOT_OK_LOG(_add_block_for_segcompaction(&block, &writer), "write block failed");
+        block.clear_column_data();
+    }
+    RETURN_NOT_OK_LOG(_check_correctness(std::move(stat), merged_row_stat, row_count, begin, end),
+                      "check correctness failed");
+    {
+        std::lock_guard<std::mutex> lock(_segid_statistics_map_mutex);
+        _clear_statistics_for_deleting_segments_unsafe(begin, end);
+    }
+    RETURN_NOT_OK(_flush_segment_writer(&writer));
+    RETURN_NOT_OK(_delete_original_segments(begin, end));
+    RETURN_NOT_OK(_rename_compacted_segments(begin, end));
+
+    if (VLOG_DEBUG_IS_ON) {
+        std::stringstream ss;
+        for (const auto& entry : std::filesystem::directory_iterator(_context.tablet_path)) {
+            ss << "[" << entry.path() << "]";
+        }
+        VLOG_DEBUG << "_segcompacted_point:" << _segcompacted_point
+                   << " _num_segment:" << _num_segment << " _num_segcompacted:" << _num_segcompacted
+                   << " list directory:" << ss.str();
+    }
+
+    _segcompacted_point += (end - begin + 1);
+
+    uint64_t elapsed = GetCurrentTimeMicros() - begin_time;
+    LOG(INFO) << "BetaRowsetWriter:" << this << " segcompaction completed. elapsed time:" << elapsed
+              << "us. _segcompacted_point update:" << _segcompacted_point;
+
+    return Status::OK();
+}
+
+void BetaRowsetWriter::segcompaction(SegCompactionCandidatesSharedPtr segments) {
+    Status status = _do_segcompaction(segments);
+    if (!status.ok()) {
+        int16_t errcode = status.precise_code();
+        switch (errcode) {
+        case OLAP_ERR_SEGCOMPACTION_INIT_READER:
+        case OLAP_ERR_SEGCOMPACTION_INIT_WRITER:
+            LOG(WARNING) << "segcompaction failed, try next time:" << status;
+            return;
+        default:
+            LOG(WARNING) << "segcompaction fatal, terminating the write job:" << status;
+            // status will be checked by the next trigger of segcompaction or the final wait
+            _segcompaction_status.store(OLAP_ERR_OTHER_ERROR);
+        }
+    }
+    DCHECK_EQ(_is_doing_segcompaction, true);
+    _is_doing_segcompaction = false;
```

Comment:
We do not hold mutex here, so the program may run into a case in which there are no thread to signal the waiting threads.